lib/SILOptimizer/FunctionSignatureTransforms/ExistentialSpecializer.cpp

```diff
@@ -0,0 +1,359 @@
+  if (!FAS)
+    return false;
+  SILInstruction *InitExistential =
+      findInitExistential(FAS.getArgumentOperands()[ArgIdx], OpenedArchetype,
+                          OpenedArchetypeDef, isCopied);
+  if (!InitExistential) {
+    LLVM_DEBUG(llvm::dbgs() << "ExistentialSpecializer Pass: Bail! Due to "
+                               "findInitExistential\n";);
+    return false;
+  }
+
+  /// Return the concrete type from init_existential returned from
+  /// findInitExistential.
+  if (auto *SingleVal = InitExistential->castToSingleValueInstruction()) {
+    if (findConcreteTypeFromInitExistential(SingleVal, ConcreteType)) {
+      return true;
+    }
+  }
+  return false;
+}
+
+/// Check if the argument Arg is used in a destroy_use instruction.
+static void
+findIfCalleeUsesArgInDestroyUse(SILValue Arg,
+                                ExistentialTransformArgumentDescriptor &ETAD) {
+  for (Operand *ArgUse : Arg->getUses()) {
+    auto *ArgUser = ArgUse->getUser();
+    if (isa<DestroyAddrInst>(ArgUser)) {
+      ETAD.DestroyAddrUse = true;
+      break;
+    }
+  }
+}
+
+/// Check if any apply argument meets the criteria for existential
+/// specialization.
+bool ExistentialSpecializer::canSpecializeExistentialArgsInFunction(
+    ApplySite &Apply,
+    llvm::SmallDenseMap<int, ExistentialTransformArgumentDescriptor>
+        &ExistentialArgDescriptor) {
+  auto *F = Apply.getReferencedFunction();
+  auto Args = F->begin()->getFunctionArguments();
+  bool returnFlag = false;
+
+  /// Analyze the argument for protocol conformance.
+  for (unsigned Idx = 0, Num = Args.size(); Idx < Num; ++Idx) {
+    auto Arg = Args[Idx];
+    auto ArgType = Arg->getType();
+    auto SwiftArgType = ArgType.getASTType();
+
+    /// Checking for AnyObject and Any is added to ensure that we do not blow up
+    /// the code size by specializing to every type that conforms to Any or
+    /// AnyObject. In future, we may want to lift these two restrictions in a
+    /// controlled way.
+    if (!ArgType.isExistentialType() || ArgType.isAnyObject() ||
+        SwiftArgType->isAny())
+      continue;
+
+    auto ExistentialRepr =
+        Arg->getType().getPreferredExistentialRepresentation(F->getModule());
+    if (!(ExistentialRepr == ExistentialRepresentation::Opaque ||
+          ExistentialRepr == ExistentialRepresentation::Class))
+      continue;
+
+    /// Find the concrete type.
+    CanType ConcreteType;
+    if (!findConcreteType(Apply, Idx, ConcreteType)) {
+      LLVM_DEBUG(
+          llvm::dbgs()
+              << "ExistentialSpecializer Pass: Bail! Due to findConcreteType "
+                 "for callee:"
+              << F->getName() << " in caller:"
+              << Apply.getInstruction()->getParent()->getParent()->getName()
+              << "\n";);
+      continue;
+    }
+
+    /// Determine attributes of the existential addr arguments such as
+    /// destroy_use, immutable_access. We should probably also disable InOut
```

Comment:
What does "probably" mean here? Either they should be disabled if they don't work, or they should be enabled